### ISSUE TITLE
[stable/keycloak] Bug report: Redirection loop when replicas > 1

### ISSUE THREAD

**Issue by swistaczek** [helm/charts#8355]
I am using chart with **Keycloak 4.5.0.Final** which is provided by keycloak chart version 4.0.0. I am trying to deploy HA version on Istio enabled cluster (1.0.2). I managed to run one replica without any troubles but when I set 2 (>1) replicas I get a problem with infinite redirections. It looks like there may be some problem with Infinispan cluster? As I understand it is used for distributed session management?

```
# keycloak-0 instance
[0m[33m06:15:58,460 WARN [org.keycloak.events] (default task-5) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.56.1.1, error=expired_code, restart_after_timeout=true

# keycloak-1 instance
[0m[33m06:15:57,974 WARN [org.keycloak.events] (default task-3) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.56.1.1, error=expired_code, restart_after_timeout=true
```

I deployed istio Gateway & Virtual Service with HTTPS in front of service provided by keycloak Helm Chart.

## Details
Chart maintainer: @unguiculus 
K8s version: 1.10.7-gke.2
I use let's encrypt ssl certificate

**Comment by swistaczek** [comment]
Sticky sessions on networking.istio.io/v1alpha3/**DestinationRule** solved issue but is sticky session required for keycloak cluster @unguiculus?

**Comment by sfitts** [comment]
They weren't in the 3.4.0 version of the chart (which uses 4.2.1.Final of Keycloak).  We only started seeing this behavior when we updated to 4.0.0.

We are actually seeing an issue with only one replica when using the following ingress:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: nginx
  labels:
    app: keycloak
    chart: keycloak-4.0.0
    heritage: Tiller
    release: keycloak
  name: keycloak
spec:
  rules:
  - http:
      paths:
      - backend:
          serviceName: keycloak-http
          servicePort: http
        path: /auth
```

This works fine using the 3.4.0 version of the chart, but in 4.0.0 we get an error when we go to the admin console -- "invalid parameter redirect_uri".  This is with a single replica.

**Comment by sfitts** [comment]
To clarify, we get the above error when we attempt to navigate to the admin console using the link on the "home page" (the one you get via /auth).

**Comment by vsomasvr** [comment]
I tried the latest version on Google Kubernetes Engine and found severe issues when replicas > 1.
(I haven't tried the previous version though. I will try it ASAP)
When replicas =1 everything is fine.
I had included all the details in my conversation here
https://github.com/helm/charts/pull/8192#issuecomment-429592731

**Comment by weisjohn** [comment]
To @sfitts point, I'm also having this issue with only one replica with an ingress controller with HTTPS. 
It was working fine when using a simple service as a LoadBalancer.

**Comment by weisjohn** [comment]
Seems like #4066 is related.

**Comment by weisjohn** [comment]
Based on the keycloak docs, I think this has everything to do with the way the helm chart is configured or rather the way that keycloak understands host names / protocols, etc.

I did find a fix, but I thought I'd document my lengthy process, maybe to give hope to others.

https://www.keycloak.org/docs/latest/server_installation/index.html#verify-configuration says to go to `https://keycloak.example.com/auth/realms/master/.well-known/openid-configuration` in order to verify the configuration. 

On my server, this looks like:

```json
{
    "issuer": "http://keycloak.example.com/auth/realms/master",
    "authorization_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/auth",
    "token_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/token",
    "token_introspection_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/token/introspect",
    "userinfo_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/userinfo",
    "end_session_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/logout",
    "jwks_uri": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/certs",
    "check_session_iframe": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/login-status-iframe.html",
    "grant_types_supported": ["authorization_code", "implicit", "refresh_token", "password", "client_credentials"],
    "response_types_supported": ["code", "none", "id_token", "token", "id_token token", "code id_token", "code token", "code id_token token"],
    "subject_types_supported": ["public", "pairwise"],
    "id_token_signing_alg_values_supported": ["ES384", "RS384", "HS256", "HS512", "ES256", "RS256", "HS384", "ES512", "RS512"],
    "userinfo_signing_alg_values_supported": ["ES384", "RS384", "HS256", "HS512", "ES256", "RS256", "HS384", "ES512", "RS512", "none"],
    "request_object_signing_alg_values_supported": ["none", "RS256"],
    "response_modes_supported": ["query", "fragment", "form_post"],
    "registration_endpoint": "http://keycloak.example.com/auth/realms/master/clients-registrations/openid-connect",
    "token_endpoint_auth_methods_supported": ["private_key_jwt", "client_secret_basic", "client_secret_post", "client_secret_jwt"],
    "token_endpoint_auth_signing_alg_values_supported": ["RS256"],
    "claims_supported": ["sub", "iss", "auth_time", "name", "given_name", "family_name", "preferred_username", "email"],
    "claim_types_supported": ["normal"],
    "claims_parameter_supported": false,
    "scopes_supported": ["openid", "phone", "address", "email", "profile", "offline_access"],
    "request_parameter_supported": true,
    "request_uri_parameter_supported": true,
    "code_challenge_methods_supported": ["plain", "S256"],
    "tls_client_certificate_bound_access_tokens": true,
    "introspection_endpoint": "http://keycloak.example.com/auth/realms/master/protocol/openid-connect/token/introspect"
}
```

You can see that there's only `http://` values in there. 

So what I seem I need to do is tell Keycloak to use HTTPS for all connections: 
https://www.keycloak.org/docs/latest/server_installation/index.html#setting-up-https-ssl

But then here I see that I need to actually use the admin to turn on SSL for the admin: https://www.keycloak.org/docs/4.5/server_admin/#_ssl_modes

So, my approach to fixing it was to use `kubectl port-forward svc/keycloak-http 8888:80` to browse to `http://localhost:8888` on my local machine to turn on SSL.

But then I hit a dead end when I refreshed the page and discovered the error message: "HTTPS required".

I need some way of telling the environment variables that I already am behind HTTPS....

So then I discovered additional environment config inside the docker specifications: https://github.com/jboss-dockerfiles/keycloak/blob/master/server/README.md#enabling-proxy-address-forwarding

I discover `PROXY_ADDRESS_FORWARDING`, which seems promising.

Here's the config that worked form me:

```yaml

## https://github.com/helm/charts/tree/master/stable/keycloak

keycloak:
  extraEnv: |
    - name: PROXY_ADDRESS_FORWARDING
      value: "true"

  persistence:
    deployPostgres: true
    dbVendor: postgres

  ## Ingress configuration.
  ## ref: https://kubernetes.io/docs/user-guide/ingress/
  ingress:
    enabled: true
    path: /

    annotations:
      kubernetes.io/ingress.class: nginx

    ## List of hosts for the ingress
    hosts:
      - keycloak.example.com

    ## TLS configuration
    tls:
      - hosts:
          - keycloak.example.com
        secretName: keycloak-ssl
```

The part that I had to enable specifically was:

```yaml
keycloak:
  extraEnv: |
    - name: PROXY_ADDRESS_FORWARDING
      value: "true"
```

Now, another issue is that trying to `helm upgrade` breaks things, but I'll save that for tomorrow.

cc: @sfitts

**Comment by vsomasvr** [comment]
@weisjohn Thanks for sharing that.
I got the single replica working with similar config as I mentioned here
https://github.com/helm/charts/pull/8192#issuecomment-429592731

Did you try with replicas > 1 ? If so, what's the behavior?
The ultimate goal is to have replication working for prod..

**Comment by weisjohn** [comment]
@vsomasvr I haven't tried with more than one replica yet.  I'm currently unable to scale the deployment at all right now.

**Comment by sfitts** [comment]
FWIW, we have the 3.4.0 chart deployed in multiple clusters, fully replicated and behind multiple ingress definitions using the following values file:

```
keycloak:
  replicas: 3

  ## Username for the initial Keycloak admin user
  username: keycloak

  ## Password for the initial Keycloak admin user
  ## If not set, a random 10 characters password will be used
  password: ""

  ## Ingress configuration.
  ## ref: https://kubernetes.io/docs/user-guide/ingress/
  ## This is the default ingress -- realm specific access is created as part of the Vantiq deploy
  ingress:
    enabled: true
    path: /auth

    annotations:
       kubernetes.io/ingress.class: nginx

    ## List of hosts for the ingress
    hosts:
      - ""

    ## TLS configuration
    tls: []
    # - hosts:
    #     - keycloak.example.com
    #   secretName: tls-keycloak
    
  persistence:

    # Disable deployment of the PostgreSQL chart
    deployPostgres: false

    # The database vendor. Can be either "postgres", "mysql", "mariadb", or "h2"
    dbVendor: postgres

    ## The following values only apply if "deployPostgres" is set to "false"

    # Optionally specify an existing secret
    existingSecret: keycloakdb
    existingSecretKey: password

    dbHost: <ourDBHost>
    dbPort: 5432 # 5432 is PostgreSQL's default port. For MySQL it would be 3306
    dbName: keycloak
    dbUser: keycloak
    dbPassword: <ourDBPassword>

  resources:
    limits:
      cpu: 1
      memory: 4Gi

    requests:
      cpu: 100m
      memory: 1Gi
```

In addition to the ingress created by the chart (which we only use for master realm access) we also deploy an ingress that looks like this (in any given cluster there may be multiple of these, each for a different v-host):

```
apiVersion: v1
items:
- apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    annotations:
      kubernetes.io/ingress.class: nginx
    name: vantiq-alik8s
    namespace: alik8s
  spec:
    rules:
    - host: alik8s.vantiq.com
      http:
        paths:
        - backend:
            serviceName: keycloak-http
            servicePort: 80
          path: /auth/
        - backend:
            serviceName: vantiq
            servicePort: 8080
          path: /
    tls:
    - hosts:
      - alik8s.vantiq.com
      secretName: vantiq-cert
```

Other details that may (or may not) matter:

* We use the Nginx Ingress controller (installed from the stable chart hosted here).
* We use a wildcard certificate for our domain
* We have deployed this on multiple Amazon EKS clusters, an Alicloud CS for K8s cluster, and an OpenStack cluster (created via Magnum).
* We've used several different PostgreSQL servers as well, so I doubt that's a factor.

All of this works perfectly using 3.4.0.

Switching to 4.0.0 while leaving everything else the same results in:

* Cluster formation failure when running replicated (exactly as reported by @vsomasvr )
* "invalid parameter" failure when running non-replicated (when accessed via the initial Ingress, we haven't tested it with the second Ingress).

Not sure if this will help anyone, but...  For now we are sticking with 3.4.0 since it is working with our configuration.

**Comment by elvinasp** [comment]
@weisjohn 

Thanks for sharing. I can confirm that on single replica install adding this `extraEnv` to Keycloak helm chart (version 4.0.0) _values.yaml_,  fixes issue with _Invalidparameter: redirect_uri_ for HTTPS ingress to access Keycloak admin interface. 

``` yaml
  keycloak:
  extraEnv: |
    - name: PROXY_ADDRESS_FORWARDING
      value: "true" 
```

Will deploy 2 node replica soon and update.

**Comment by vsomasvr** [comment]
@sfitts 
I guess 

> Nginx Ingress controller (installed from the stable chart hosted here).

would make a lot of difference. I will try this today
Thank you very much for sharing this

**Comment by sfitts** [comment]
@elvinasp thanks for the update on that.

Based on the comments in #8192 it seems that the cluster issue may be related to the switch from JDBC_PING to DNS_PING.  Has anyone tried switching back?  Haven't looked at this in detail yet.

**Comment by monotek** [comment]
@axdotl is this the problem you're also facing?

**Comment by vsomasvr** [comment]
I had tried it (version: 3.4.0) using nginx ingress  (as mentioned by @sfitts) , and it worked on GKE, great to have keycloak cluster deployed successfully..

**Comment by vsomasvr** [comment]
I tried deploying the latest version 4.0.0 kc helm chart (with 4.5.0 Final keycloak) with JDBC_PING settings from 3.4.0 kc helm chart (with 4.2.1 Final keycloak) and found issues with the deployment. 
Clearly, the JGroups JDBC_PING settings from 3.4.0 version don't work for the 4.0.0 version.

I had posted this question on stackoverflow here
https://stackoverflow.com/questions/52901704/keycloak-4-5-0-final-cluster-deployment-jdbc-ping-cli-config

**Comment by jreyes** [comment]
> ```
> # keycloak-0 instance
> [0m�[33m06:15:58,460 WARN [org.keycloak.events] (default task-5) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.56.1.1, error=expired_code, restart_after_timeout=true
> 
> # keycloak-1 instance
> [0m�[33m06:15:57,974 WARN [org.keycloak.events] (default task-3) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.56.1.1, error=expired_code, restart_after_timeout=true
> ```

@swistaczek I'm getting the same expired_code error in a deployment that used to work before the update to 4.5.0.Final. This is in a clean installation and trying to login in the Admin console. In my case I'm not using Istio and using 2 replicas. Does anyone else was able to resolve it?

**Comment by styk-tv** [comment]
I'm hitting the same on kube with haproxy ingress, on aws and postrgres in rds (this bit doesn't matter most likely. I found this excerpt from documentation ref https://www.keycloak.org/docs/4.5/server_installation/#troubleshooting

> If you try to authenticate with Keycloak to your application, but authentication fails with an infinite number of redirects in your browser and you see the errors like this in the Keycloak server log:

```
2017-11-27 14:50:31,587 WARN  [org.keycloak.events] (default task-17) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=aa.bb.cc.dd, error=expired_code, restart_after_timeout=true
```

> it probably means that your load balancer needs to be set to support sticky sessions. Make sure that the provided route name used during startup of Keycloak server (Property jboss.node.name) contains the correct name used by the load balancer server to identify the current server.

**Comment by mborst** [comment]
This might be related: https://github.com/helm/charts/issues/7679#issuecomment-430254416

**Comment by knutjda** [comment]
Rebalancing and initial discovery works when I added that line to keycloak.cli.ha.

**Comment by vsomasvr** [comment]
> I opened a PR for this but in the meantime you should be able to add this line to your keycloak.cli.custom or ha section: `/subsystem=jgroups/channel=ee:write-attribute(name=stack, value=tcp)` to force jgroups to use tcp instead of udp

Cool, this has worked! Keycloak was able to start in cluster mode successfully..
I tried this on GKE, worked as expected with gce loadbalancer

**Comment by svyotov** [comment]
I had the same issue.  Sticky sessions and enabling proxy as extra env var resolved it.

From the official documentation https://www.keycloak.org/docs/3.4/server_installation : 

```
If you try to authenticate with Keycloak to your application, but it failed with the infinite number of redirects in your browser and you see the errors like this in the Keycloak server log:

2017-11-27 14:50:31,587 WARN  [org.keycloak.events] (default task-17) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=aa.bb.cc.dd, error=expired_code, restart_after_timeout=true
it probably means that your load balancer needs to be set to support sticky sessions. Make sure that the provided route name used during startup of Keycloak server (Property jboss.node.name) contains the correct name used by the load balancer server to identify the current server.
```

TL;DR
---------
The ingress/ load balancer/ proxy configuration needs to support sticky sessions.

**Comment by dojadop** [comment]
Sticky sessions shouldn't be required and we are not using them in our deployment behind nginx-ingress.  

The sticky session is not mandatory for the cluster setup, however it is good for performance for the reasons mentioned above. https://www.keycloak.org/docs/3.4/server_installation/#sticky-sessions

The proxy_address_forwarding is absolutely required in the latest version of keycloak (I believe this is because the intenal keycloak config file changed) when you are behind an ingress with tls so the server knows to convert http to https.

**Comment by mborst** [comment]
I have got it to run without `PROXY_ADDRESS_FORWARDING`, see https://github.com/helm/charts/issues/7679#issuecomment-430254416. IMHO, the reason is a mismatch between this chart and the keycloak image.

**Comment by dojadop** [comment]
Thanks @mborst I think I see now why it works based on your comment if you include the standalone.xml.  Looks like #8777 should fix this issue without having to include the extraArgs. As for the Ha configuration I think I have a solution that works without having to include the PROXY_ADDRESS_FORWARDING and will open a PR for that

**Comment by jam13** [comment]
Still seeing the redirect loop problem with v4.0.4 of the chart. Works ok with a single replica, but 2+ fails to set up infinispan correctly.

Setup:
- Google Kubernetes (two node cluster)
- Traefik as ingress handler with L4 gcloud load balancer in front
- PROXY_ADDRESS_FORWARDING enabled

**Comment by jam13** [comment]
SOLVED.

I'm using pulumi to deploy the helm chart and for some reason it was deploying to `default` but `.Release.Namespace` was set to `platform` which meant it was using the wrong hostname for DNS_PING (`keycloak-headless.platform.svc.cluster.local`). Explicitly defining the namespace as `default` fixed it and it's now working correctly.

I think this is related to pulumi's handling of namespaces ref: https://github.com/helm/helm/issues/3553

**Comment by Ticiano-mw** [comment]
I solved this by making sure my JGROUP_ env vars for the KEycloak docker image were correctly set. As this manages the service discovery + Clustering via "Infinispan".

There are two (actually 3 counting JDBC) discovery options, in K8s:  DNS_PING or KUBE_PING

if using DNS_PING you need to set the "property" to the proper DNS record of the service  (considering namespace and cluster domain), e.g.:

```
        - name: JGROUPS_DISCOVERY_PROTOCOL
          value: dns.DNS_PING
        - name: JGROUPS_DISCOVERY_PROPERTIES
          value: dns_query=keycloak-discovery.security
        - name: CACHE_OWNERS_COUNT
          value: "2"
        - name: CACHE_OWNERS_AUTH_SESSIONS_COUNT
          value: "2"
```

Note these are env variables passed to the Keycloak container (the  CACHE_OWNERS_*  are always present and need to be >=2 for HA)

in the case of using KUBE_PING, it looks like this:

```
            - name: JGROUPS_DISCOVERY_PROTOCOL
              value: kubernetes.KUBE_PING
            - name: JGROUPS_DISCOVERY_PROPERTIES
              value: 'namespace=${KUBERNETES_NAMESPACE:<your_namespace_name>}'
            - name: JGROUPS_TRANSPORT_STACK
              value: tcp
            - name: CACHE_OWNERS_COUNT
              value: '2'
            - name: AUTH_CACHE_OWNERS_COUNT
              value: '2'
```

or even:

```
            - name: JGROUPS_DISCOVERY_PROTOCOL
              value: kubernetes.KUBE_PING
            - name: KUBERNETES_NAMESPACE
              value: <your_namespace_name>
            - name: CACHE_OWNERS_COUNT
              value: '2'
            - name: AUTH_CACHE_OWNERS_COUNT
              value: '2'
```

`JGROUPS_TRANSPORT_STACK` seems to be optional (probably defaults to tcp), and the  `KUBERNETES_NAMESPACE` is actually read by Jgroups to figure the ns out.

**Comment by claudiohs-inditex** [comment]
I have the same issue and find a solution without the use of localhost or 127.0.01
I put the [solution here](https://github.com/Apicurio/apicurio-registry/issues/2119#issuecomment-997845187)
